<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Variant Benchmarking Tools (VBT): CVariant Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Variant Benchmarking Tools (VBT)
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="class_c_variant-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">CVariant Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>A Container that stores the necessary information of variant.  
 <a href="class_c_variant.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="_c_variant_8h_source.html">CVariant.h</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a8dec728411cc0d31d29b2475146d298b"><td class="memItemLeft" align="right" valign="top"><a id="a8dec728411cc0d31d29b2475146d298b"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a8dec728411cc0d31d29b2475146d298b">CVariant</a> ()</td></tr>
<tr class="memdesc:a8dec728411cc0d31d29b2475146d298b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default constructor. <br /></td></tr>
<tr class="separator:a8dec728411cc0d31d29b2475146d298b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5a0ac7949809ed3e2745abb3ce3d312d"><td class="memItemLeft" align="right" valign="top"><a id="a5a0ac7949809ed3e2745abb3ce3d312d"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a5a0ac7949809ed3e2745abb3ce3d312d">CVariant</a> (const <a class="el" href="class_c_variant.html">CVariant</a> &amp;a_rObj)</td></tr>
<tr class="memdesc:a5a0ac7949809ed3e2745abb3ce3d312d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Copy constructor. <br /></td></tr>
<tr class="separator:a5a0ac7949809ed3e2745abb3ce3d312d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6db18efbcd5e646e5a81ce537d0c7963"><td class="memItemLeft" align="right" valign="top"><a id="a6db18efbcd5e646e5a81ce537d0c7963"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a6db18efbcd5e646e5a81ce537d0c7963">Clear</a> ()</td></tr>
<tr class="memdesc:a6db18efbcd5e646e5a81ce537d0c7963"><td class="mdescLeft">&#160;</td><td class="mdescRight">Clear the Variant. <br /></td></tr>
<tr class="separator:a6db18efbcd5e646e5a81ce537d0c7963"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2cb7cf05e9f0466a3fe1e70f724460d9"><td class="memItemLeft" align="right" valign="top"><a id="a2cb7cf05e9f0466a3fe1e70f724460d9"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a2cb7cf05e9f0466a3fe1e70f724460d9">CompareTo</a> (const <a class="el" href="class_c_variant.html">CVariant</a> &amp;a_rObj) const</td></tr>
<tr class="memdesc:a2cb7cf05e9f0466a3fe1e70f724460d9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compare the given variant with this. <br /></td></tr>
<tr class="separator:a2cb7cf05e9f0466a3fe1e70f724460d9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae604c4b9ad5adb6d2212a9f93d44b539"><td class="memItemLeft" align="right" valign="top"><a id="ae604c4b9ad5adb6d2212a9f93d44b539"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#ae604c4b9ad5adb6d2212a9f93d44b539">IsPhased</a> () const</td></tr>
<tr class="memdesc:ae604c4b9ad5adb6d2212a9f93d44b539"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return if genotype is phased. <br /></td></tr>
<tr class="separator:ae604c4b9ad5adb6d2212a9f93d44b539"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae5e33ee8ee849a4b2ca0f0780c8e7bbb"><td class="memItemLeft" align="right" valign="top"><a id="ae5e33ee8ee849a4b2ca0f0780c8e7bbb"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#ae5e33ee8ee849a4b2ca0f0780c8e7bbb">GetStart</a> () const</td></tr>
<tr class="memdesc:ae5e33ee8ee849a4b2ca0f0780c8e7bbb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the start position of the Variant. <br /></td></tr>
<tr class="separator:ae5e33ee8ee849a4b2ca0f0780c8e7bbb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac71c09ba7acb6edc3364db8cd46b4c06"><td class="memItemLeft" align="right" valign="top"><a id="ac71c09ba7acb6edc3364db8cd46b4c06"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#ac71c09ba7acb6edc3364db8cd46b4c06">GetEnd</a> () const</td></tr>
<tr class="memdesc:ac71c09ba7acb6edc3364db8cd46b4c06"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the end position of the ref allele. <br /></td></tr>
<tr class="separator:ac71c09ba7acb6edc3364db8cd46b4c06"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab52d67d665f71b0947641a5229ec5edd"><td class="memItemLeft" align="right" valign="top"><a id="ab52d67d665f71b0947641a5229ec5edd"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#ab52d67d665f71b0947641a5229ec5edd">GetOriginalPos</a> () const</td></tr>
<tr class="memdesc:ab52d67d665f71b0947641a5229ec5edd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the original position of the variant ignoring trimming and first nucleotide reduction. <br /></td></tr>
<tr class="separator:ab52d67d665f71b0947641a5229ec5edd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2e68348e9a50458bccd7c863a4c3269a"><td class="memItemLeft" align="right" valign="top"><a id="a2e68348e9a50458bccd7c863a4c3269a"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a2e68348e9a50458bccd7c863a4c3269a">IsHeterozygous</a> () const</td></tr>
<tr class="memdesc:a2e68348e9a50458bccd7c863a4c3269a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return is the variant is heterozygous. <br /></td></tr>
<tr class="separator:a2e68348e9a50458bccd7c863a4c3269a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1018380af3b5ad539e827e4c66733e1b"><td class="memItemLeft" align="right" valign="top"><a id="a1018380af3b5ad539e827e4c66733e1b"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a1018380af3b5ad539e827e4c66733e1b">Print</a> () const</td></tr>
<tr class="memdesc:a1018380af3b5ad539e827e4c66733e1b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the Variant. <br /></td></tr>
<tr class="separator:a1018380af3b5ad539e827e4c66733e1b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a32f32f487dc4e79483fdb3240e932d43"><td class="memItemLeft" align="right" valign="top"><a id="a32f32f487dc4e79483fdb3240e932d43"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a32f32f487dc4e79483fdb3240e932d43">IsNull</a> () const</td></tr>
<tr class="memdesc:a32f32f487dc4e79483fdb3240e932d43"><td class="mdescLeft">&#160;</td><td class="mdescRight">Check if variant is initialized. <br /></td></tr>
<tr class="separator:a32f32f487dc4e79483fdb3240e932d43"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1d47eae2bd731693f555af5818f168a7"><td class="memItemLeft" align="right" valign="top"><a id="a1d47eae2bd731693f555af5818f168a7"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a1d47eae2bd731693f555af5818f168a7">GetId</a> () const</td></tr>
<tr class="memdesc:a1d47eae2bd731693f555af5818f168a7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the unique id of the variant. <br /></td></tr>
<tr class="separator:a1d47eae2bd731693f555af5818f168a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6b8854be5475ca0857a5f3fa740bd18d"><td class="memItemLeft" align="right" valign="top"><a id="a6b8854be5475ca0857a5f3fa740bd18d"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a6b8854be5475ca0857a5f3fa740bd18d">IsFilterPASS</a> () const</td></tr>
<tr class="memdesc:a6b8854be5475ca0857a5f3fa740bd18d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return true if the variant filter column is PASS. <br /></td></tr>
<tr class="separator:a6b8854be5475ca0857a5f3fa740bd18d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae735e75c01402d9ee15b9ddf005c5557"><td class="memItemLeft" align="right" valign="top"><a id="ae735e75c01402d9ee15b9ddf005c5557"></a>
std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#ae735e75c01402d9ee15b9ddf005c5557">GetOriginalAlleleStr</a> (unsigned int a_nAlleleIndex) const</td></tr>
<tr class="memdesc:ae735e75c01402d9ee15b9ddf005c5557"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the original Alt string with the given index (Take back trim operation) <br /></td></tr>
<tr class="separator:ae735e75c01402d9ee15b9ddf005c5557"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a214fbfed5529e0f7189c5e4e3cbc1614"><td class="memItemLeft" align="right" valign="top"><a id="a214fbfed5529e0f7189c5e4e3cbc1614"></a>
std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a214fbfed5529e0f7189c5e4e3cbc1614">GetRefSeq</a> () const</td></tr>
<tr class="memdesc:a214fbfed5529e0f7189c5e4e3cbc1614"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the reference sequences. <br /></td></tr>
<tr class="separator:a214fbfed5529e0f7189c5e4e3cbc1614"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aab61403bf9dbc4255681924af20bdbb3"><td class="memItemLeft" align="right" valign="top"><a id="aab61403bf9dbc4255681924af20bdbb3"></a>
<a class="el" href="struct_s_allele.html">SAllele</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#aab61403bf9dbc4255681924af20bdbb3">GetAllele</a> (int a_nAlleleId) const</td></tr>
<tr class="memdesc:aab61403bf9dbc4255681924af20bdbb3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the allele sequence specified with the id (0 is first allele, 1 is second allele) <br /></td></tr>
<tr class="separator:aab61403bf9dbc4255681924af20bdbb3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5bf282c0f3fdd73eb22cb773be6dd5b3"><td class="memItemLeft" align="right" valign="top"><a id="a5bf282c0f3fdd73eb22cb773be6dd5b3"></a>
EVariantType&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a5bf282c0f3fdd73eb22cb773be6dd5b3">GetVariantType</a> () const</td></tr>
<tr class="memdesc:a5bf282c0f3fdd73eb22cb773be6dd5b3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the type of variant. <br /></td></tr>
<tr class="separator:a5bf282c0f3fdd73eb22cb773be6dd5b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae141c863c736f12302bd8b9f24c7b4c2"><td class="memItemLeft" align="right" valign="top"><a id="ae141c863c736f12302bd8b9f24c7b4c2"></a>
EVariantCategory&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#ae141c863c736f12302bd8b9f24c7b4c2">GetVariantCategory</a> () const</td></tr>
<tr class="memdesc:ae141c863c736f12302bd8b9f24c7b4c2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the category of variant to use in mendelian violation reporting. <br /></td></tr>
<tr class="separator:ae141c863c736f12302bd8b9f24c7b4c2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a58e053132a175be80541f968ec4a91d6"><td class="memItemLeft" align="right" valign="top"><a id="a58e053132a175be80541f968ec4a91d6"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a58e053132a175be80541f968ec4a91d6">GetGenotypeArr</a> (int *a_pGenotypeList, int &amp;a_rGenotypeCount)</td></tr>
<tr class="memdesc:a58e053132a175be80541f968ec4a91d6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Fill the Genotype list with the original genotype indexes and the genotype count. <br /></td></tr>
<tr class="separator:a58e053132a175be80541f968ec4a91d6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a69b6142175c739399467dd0969a1db3a"><td class="memItemLeft" align="right" valign="top"><a id="a69b6142175c739399467dd0969a1db3a"></a>
std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a69b6142175c739399467dd0969a1db3a">ToString</a> () const</td></tr>
<tr class="memdesc:a69b6142175c739399467dd0969a1db3a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print the variant [For Test Purpose]. <br /></td></tr>
<tr class="separator:a69b6142175c739399467dd0969a1db3a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a06cab98a71ae09d7911a74dae9bdd248"><td class="memItemLeft" align="right" valign="top"><a id="a06cab98a71ae09d7911a74dae9bdd248"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a06cab98a71ae09d7911a74dae9bdd248">GetMaxTrimStartEnd</a> (int a_nAlleleIndex, unsigned int &amp;trimLengthFromBeginning, unsigned int &amp;trimLengthFromEnd)</td></tr>
<tr class="memdesc:a06cab98a71ae09d7911a74dae9bdd248"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets the maximum number of nucleotides can be trimmed from beginning and ending of selected allele. <br /></td></tr>
<tr class="separator:a06cab98a71ae09d7911a74dae9bdd248"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5415fc6f7ea4d01de323d918066721f0"><td class="memItemLeft" align="right" valign="top"><a id="a5415fc6f7ea4d01de323d918066721f0"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a5415fc6f7ea4d01de323d918066721f0">TrimVariant</a> (int a_nAlleleIndex)</td></tr>
<tr class="memdesc:a5415fc6f7ea4d01de323d918066721f0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Trim the redundant nucleotides from beginning and ending of given allele (Beginning nucleotides will be trimmed first) <br /></td></tr>
<tr class="separator:a5415fc6f7ea4d01de323d918066721f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3ddba7df62d8bc15730fa4123893c9e3"><td class="memItemLeft" align="right" valign="top"><a id="a3ddba7df62d8bc15730fa4123893c9e3"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a3ddba7df62d8bc15730fa4123893c9e3">TrimVariant</a> (int a_nAlleleIndex, unsigned int trimLengthFromBeginning, unsigned int trimLengthFromEnd)</td></tr>
<tr class="memdesc:a3ddba7df62d8bc15730fa4123893c9e3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Trim the redundant nucleotides from beginning and ending of given allele. Nucleotides to be clipped from beginning and ending of the allele are specified as parameter. <br /></td></tr>
<tr class="separator:a3ddba7df62d8bc15730fa4123893c9e3"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a5fb800f342a8ce9756e30f00d09220a3"><td class="memItemLeft" align="right" valign="top"><a id="a5fb800f342a8ce9756e30f00d09220a3"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a5fb800f342a8ce9756e30f00d09220a3">m_nVcfId</a></td></tr>
<tr class="memdesc:a5fb800f342a8ce9756e30f00d09220a3"><td class="mdescLeft">&#160;</td><td class="mdescRight">ID of which vcf file that the variant belongs to. <br /></td></tr>
<tr class="separator:a5fb800f342a8ce9756e30f00d09220a3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6d0ced08dcc1d4710e871d4a09eae50f"><td class="memItemLeft" align="right" valign="top"><a id="a6d0ced08dcc1d4710e871d4a09eae50f"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a6d0ced08dcc1d4710e871d4a09eae50f">m_nChrId</a></td></tr>
<tr class="memdesc:a6d0ced08dcc1d4710e871d4a09eae50f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Id of the chromosome that variant belogs to. <br /></td></tr>
<tr class="separator:a6d0ced08dcc1d4710e871d4a09eae50f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5b05f7e9c5af0753de7b69ac2553ae15"><td class="memItemLeft" align="right" valign="top"><a id="a5b05f7e9c5af0753de7b69ac2553ae15"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a5b05f7e9c5af0753de7b69ac2553ae15">m_nId</a></td></tr>
<tr class="memdesc:a5b05f7e9c5af0753de7b69ac2553ae15"><td class="mdescLeft">&#160;</td><td class="mdescRight">Unique Id of variant. <br /></td></tr>
<tr class="separator:a5b05f7e9c5af0753de7b69ac2553ae15"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a51214c8fd9b759aaacf7143290e0c88e"><td class="memItemLeft" align="right" valign="top"><a id="a51214c8fd9b759aaacf7143290e0c88e"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a51214c8fd9b759aaacf7143290e0c88e">m_nAlleleCount</a></td></tr>
<tr class="memdesc:a51214c8fd9b759aaacf7143290e0c88e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allele count of the variant (2 for diploid and 1 for haploid) <br /></td></tr>
<tr class="separator:a51214c8fd9b759aaacf7143290e0c88e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa70c38a0bec893b2fc3c8e93842ef711"><td class="memItemLeft" align="right" valign="top"><a id="aa70c38a0bec893b2fc3c8e93842ef711"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#aa70c38a0bec893b2fc3c8e93842ef711">m_nStartPos</a></td></tr>
<tr class="memdesc:aa70c38a0bec893b2fc3c8e93842ef711"><td class="mdescLeft">&#160;</td><td class="mdescRight">Start Position of the variant - min start pos of all alleles. <br /></td></tr>
<tr class="separator:aa70c38a0bec893b2fc3c8e93842ef711"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3336b7e5a09b9fa87c6f739c8595cf12"><td class="memItemLeft" align="right" valign="top"><a id="a3336b7e5a09b9fa87c6f739c8595cf12"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a3336b7e5a09b9fa87c6f739c8595cf12">m_nEndPos</a></td></tr>
<tr class="memdesc:a3336b7e5a09b9fa87c6f739c8595cf12"><td class="mdescLeft">&#160;</td><td class="mdescRight">End Position of the variant - max end pos of all alleles. <br /></td></tr>
<tr class="separator:a3336b7e5a09b9fa87c6f739c8595cf12"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5db1d049728f9ccb866cf26508d3b88e"><td class="memItemLeft" align="right" valign="top"><a id="a5db1d049728f9ccb866cf26508d3b88e"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a5db1d049728f9ccb866cf26508d3b88e">m_genotype</a> [2]</td></tr>
<tr class="memdesc:a5db1d049728f9ccb866cf26508d3b88e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Original Genotype list read from vcf file. <br /></td></tr>
<tr class="separator:a5db1d049728f9ccb866cf26508d3b88e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9fda400457edd454c265380e3d4fdf8e"><td class="memItemLeft" align="right" valign="top"><a id="a9fda400457edd454c265380e3d4fdf8e"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a9fda400457edd454c265380e3d4fdf8e">m_nZygotCount</a></td></tr>
<tr class="memdesc:a9fda400457edd454c265380e3d4fdf8e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Zygot Count. <br /></td></tr>
<tr class="separator:a9fda400457edd454c265380e3d4fdf8e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a30862cd7015d9d3f267d94333075e429"><td class="memItemLeft" align="right" valign="top"><a id="a30862cd7015d9d3f267d94333075e429"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a30862cd7015d9d3f267d94333075e429">m_nOriginalPos</a></td></tr>
<tr class="memdesc:a30862cd7015d9d3f267d94333075e429"><td class="mdescLeft">&#160;</td><td class="mdescRight">Original variant position. <br /></td></tr>
<tr class="separator:a30862cd7015d9d3f267d94333075e429"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a291c56976d16636d1622173a65444344"><td class="memItemLeft" align="right" valign="top"><a id="a291c56976d16636d1622173a65444344"></a>
EVariantMatch&#160;</td><td class="memItemRight" valign="bottom"><b>m_variantStatus</b></td></tr>
<tr class="separator:a291c56976d16636d1622173a65444344"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a87c3d2e3899506aaa242fe9dd968c58c"><td class="memItemLeft" align="right" valign="top"><a id="a87c3d2e3899506aaa242fe9dd968c58c"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a87c3d2e3899506aaa242fe9dd968c58c">m_bIsPhased</a></td></tr>
<tr class="memdesc:a87c3d2e3899506aaa242fe9dd968c58c"><td class="mdescLeft">&#160;</td><td class="mdescRight">True if the variant genotype is phased. <br /></td></tr>
<tr class="separator:a87c3d2e3899506aaa242fe9dd968c58c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a584b4e62c45f2d173da5fc8e224655a2"><td class="memItemLeft" align="right" valign="top"><a id="a584b4e62c45f2d173da5fc8e224655a2"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a584b4e62c45f2d173da5fc8e224655a2">m_bIsHeterozygous</a></td></tr>
<tr class="memdesc:a584b4e62c45f2d173da5fc8e224655a2"><td class="mdescLeft">&#160;</td><td class="mdescRight">If the variant is heterozygous (ie. GT is 0/1 1/2 etc.) <br /></td></tr>
<tr class="separator:a584b4e62c45f2d173da5fc8e224655a2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af8a629628e2960a4f6aa416abe468dd3"><td class="memItemLeft" align="right" valign="top"><a id="af8a629628e2960a4f6aa416abe468dd3"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#af8a629628e2960a4f6aa416abe468dd3">m_bIsFirstNucleotideTrimmed</a></td></tr>
<tr class="memdesc:af8a629628e2960a4f6aa416abe468dd3"><td class="mdescLeft">&#160;</td><td class="mdescRight">True if the first nucleotide is trimmed. <br /></td></tr>
<tr class="separator:af8a629628e2960a4f6aa416abe468dd3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a54274d28ccf03a181f3a1f3f64668309"><td class="memItemLeft" align="right" valign="top"><a id="a54274d28ccf03a181f3a1f3f64668309"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a54274d28ccf03a181f3a1f3f64668309">m_bIsNoCall</a></td></tr>
<tr class="memdesc:a54274d28ccf03a181f3a1f3f64668309"><td class="mdescLeft">&#160;</td><td class="mdescRight">True if genotype is ./. <br /></td></tr>
<tr class="separator:a54274d28ccf03a181f3a1f3f64668309"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a091c9b95b494b22b12732666ae731c84"><td class="memItemLeft" align="right" valign="top"><a id="a091c9b95b494b22b12732666ae731c84"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a091c9b95b494b22b12732666ae731c84">m_bHaveMultipleTrimOption</a></td></tr>
<tr class="memdesc:a091c9b95b494b22b12732666ae731c84"><td class="mdescLeft">&#160;</td><td class="mdescRight">If the variant can be trimmed more than 1 way (for -ref-overlap mode) <br /></td></tr>
<tr class="separator:a091c9b95b494b22b12732666ae731c84"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aabc06e2661a97fe9aaaa1de67459fdd1"><td class="memItemLeft" align="right" valign="top"><a id="aabc06e2661a97fe9aaaa1de67459fdd1"></a>
<a class="el" href="struct_s_allele.html">SAllele</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#aabc06e2661a97fe9aaaa1de67459fdd1">m_alleles</a> [2]</td></tr>
<tr class="memdesc:aabc06e2661a97fe9aaaa1de67459fdd1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Allele array of the variant. <br /></td></tr>
<tr class="separator:aabc06e2661a97fe9aaaa1de67459fdd1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5b3c5b7c0d145b2bc7338ecdd2d313f2"><td class="memItemLeft" align="right" valign="top"><a id="a5b3c5b7c0d145b2bc7338ecdd2d313f2"></a>
std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a5b3c5b7c0d145b2bc7338ecdd2d313f2">m_refSequence</a></td></tr>
<tr class="memdesc:a5b3c5b7c0d145b2bc7338ecdd2d313f2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Reference sequence. <br /></td></tr>
<tr class="separator:a5b3c5b7c0d145b2bc7338ecdd2d313f2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a26cdaefec2b7831f09e9a1d43e9a3234"><td class="memItemLeft" align="right" valign="top"><a id="a26cdaefec2b7831f09e9a1d43e9a3234"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a26cdaefec2b7831f09e9a1d43e9a3234">m_bIsFilterPASS</a></td></tr>
<tr class="memdesc:a26cdaefec2b7831f09e9a1d43e9a3234"><td class="mdescLeft">&#160;</td><td class="mdescRight">Filter Data. <br /></td></tr>
<tr class="separator:a26cdaefec2b7831f09e9a1d43e9a3234"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4e092539e4e5ef1a5f11f3a8bd6de1f3"><td class="memItemLeft" align="right" valign="top"><a id="a4e092539e4e5ef1a5f11f3a8bd6de1f3"></a>
std::vector&lt; std::string &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>m_filterString</b></td></tr>
<tr class="separator:a4e092539e4e5ef1a5f11f3a8bd6de1f3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab855bdcfb96e2d3c07b6e08ce8268d15"><td class="memItemLeft" align="right" valign="top"><a id="ab855bdcfb96e2d3c07b6e08ce8268d15"></a>
std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#ab855bdcfb96e2d3c07b6e08ce8268d15">m_chrName</a></td></tr>
<tr class="memdesc:ab855bdcfb96e2d3c07b6e08ce8268d15"><td class="mdescLeft">&#160;</td><td class="mdescRight">Chromosome name. <br /></td></tr>
<tr class="separator:ab855bdcfb96e2d3c07b6e08ce8268d15"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a609fff1b972280c7e413e668991dfe2a"><td class="memItemLeft" align="right" valign="top"><a id="a609fff1b972280c7e413e668991dfe2a"></a>
std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_c_variant.html#a609fff1b972280c7e413e668991dfe2a">m_allelesStr</a></td></tr>
<tr class="memdesc:a609fff1b972280c7e413e668991dfe2a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Original Alleles string read from vcf file. <br /></td></tr>
<tr class="separator:a609fff1b972280c7e413e668991dfe2a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>A Container that stores the necessary information of variant. </p>
<p><a class="el" href="class_c_variant.html" title="A Container that stores the necessary information of variant. ">CVariant</a> is a container for variants that stores necessary vcfrecord information read with htslib </p>
</div><hr/>The documentation for this class was generated from the following files:<ul>
<li>/Users/c1ms21p6h3qk/Desktop/VBT/vbt/VcfIO/include/<a class="el" href="_c_variant_8h_source.html">CVariant.h</a></li>
<li>/Users/c1ms21p6h3qk/Desktop/VBT/vbt/VcfIO/src/CVariant.cpp</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.13
</small></address>
</body>
</html>
